design/pipeline-caching.md

```diff
@@ -0,0 +1,157 @@
+
+```yaml
+steps:
+- restore_cache: yarn.lock
+- script: yarn
+- save_cache:
+  keys: yarn.lock
+  paths: node_modules
+- script: yarn build
+```
+
+This more compact syntax just reduces some of the noise around using the caching task. Additionally we are planning to integrate caching into the the ```use:``` syntax. This will look like the following:
+
+```yaml
+steps:
+- use: node
+  cache: true
+- script: |
+    yarn
+    yarn build
+    yarn test
+```
+
+The idea behind the ```use:``` syntax is that given what we are using (in this case ```node```) we make some automatic decisions about what and how to cache. This logic will use various heuristics to make the decision about what to cache, for example, it will look for the presence of a ```yarn.lock``` file or a ```package-lock.json``` to determine where the ```node_modules``` path may be located.
+
+Other ```use:``` statements will have similar heuristics to pick the best defaults.
+
+Note that the ```use:``` syntax will inject the cache save step at the end of the build process which will not always be desirable. In those cases developers can fall back to specifying the ```restore_cache:``` and ```save_cache:``` YAML statements.
+
+### Cache Scoping
+
+Getting scoping right is important for maximising cache hits and but also avoiding the cache becoming an attack vector to insert malicous code into official/master builds.
+
+As a result we will automatically scope caches to the branch that they are running against. The caches will also be hierarchical so a feature branch will be able to get a hit on the master, but when populatingn the cache in the ```save_cache:``` step, the contents of that cache won't be used on the master build.
+
+For PR builds, the PR build will use the cache of the branch it is merging into or from with preference been given to the branch it is merging from. Once again, the PR build itself will not be able to add content to the scope of the branch being merged into to avoid cache attack vectors.
+
+### Cache Expiry
+
+Cache expiry will be at a minimum, 7-days. However we will use hit counters to keep cache entries alive and may extend the minimum retention to a longer period if there are projects that build less frequently that could benefit from longer lifetimes.
+
+### Step Over Support
+
+In the interests of correctness we will not skip build steps by default based on a cache hit. However we will provide a way to emit a variable which can be used in subsequent steps to skip a task if a cache is hit. The usage is as follows:
+
+```yaml
+steps:
+- restore_cache: yarn.lock
+  skip-variable: cache.skipyarn
+- script: yarn
+  condition: eq(variables['cache.skipyarn'], 'sourcehit')
+```
+
+The value inserted into the variable specified by ```skip-variable:``` will change depending on whether there was a cache hit or miss, and what kind of hit it was. Example values are:
+
+* sourcehit; used to signal that there is a cache hit on the source branch in a PR.
```

Comment:
Hmm, I didn't follow here.  Up to this point, a hit is a hit is a hit - doesn't matter where it came from or how it got there as long as the key matches.

Reply:
Initially I was thinking a true/false value. However upon reflection I decided that it might be better to know what scope the cache hit came from because you might decide that if I got a cache hit based on the source branch then I could do a step over (because that cache is likely to be accurate and not be missing any dependencies). However - if it was a hit on the target branch (in the case of a PR build) then you might want to do the step because master is less likely to have any missing dependencies (assuming you've added some).

In the simple case you would just check to make sure it has a non-empty string to do step over. But once again step over wouldn't be a default behavior - its an optimization that you go into knowing the issues :)

Reply:
I'm not clear on how this works when the key is the same across branches. Does the first (the PR branch) always win? Do we store a set of all branches?

Reply:
They key might be the same across branches, but because it is in a different scope it shouldn't matter. You could have the same key in different scopes pointing to the same, or different content (although I think it would typically be the same).

Reply:
Here is the logic that I would apply:

1. Try to restore cache from the source branch scope (regardless of whether it is a PR or branch build)
2. If #1 fails and is a PR try to restore cache from target branch scope.
3. If #2 fails and is not a PR try to restore from parent branch scope.

Is it a good idea to explicitly assume Git here?

Reply:
Ah ok, so there's a namespacing that's happening and it's still reasonably efficient because of deduplication.